core/drivers/bcm_gpio.c

```diff
@@ -0,0 +1,175 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright 2019 Broadcom.
+ */
+#include <assert.h>
+#include <drivers/bcm_gpio.h>
+#include <initcall.h>
+#include <io.h>
+#include <mm/core_memprot.h>
+#include <platform_config.h>
+#include <trace.h>
+
+#define IPROC_GPIO_DATA_IN_OFFSET	0x00
+#define IPROC_GPIO_DATA_OUT_OFFSET	0x04
+#define IPROC_GPIO_OUT_EN_OFFSET	0x08
+#define IPROC_GPIO_INT_MSK_OFFSET	0x18
+
+#define GPIO_BANK_SIZE			0x200
+#define NGPIOS_PER_BANK			32
+#define GPIO_BANK(pin)			((pin) / NGPIOS_PER_BANK)
+
+#define IPROC_GPIO_REG(pin, reg)	((reg) + \
+	GPIO_BANK(pin) * GPIO_BANK_SIZE)
+
+#define IPROC_GPIO_SHIFT(pin)		((pin) % NGPIOS_PER_BANK)
+
+static SLIST_HEAD(, bcm_gpio_chip) gclist = SLIST_HEAD_INITIALIZER(gclist);
+
+struct bcm_gpio_chip *bcm_gpio_pin_to_chip(unsigned int pin)
+{
+	struct bcm_gpio_chip *gc = NULL;
+
+	SLIST_FOREACH(gc, &gclist, link)
+		if ((pin >= gc->gpio_base) &&
+		    (pin < (gc->gpio_base + gc->ngpios)))
+			return gc;
+	return NULL;
+}
+
+static bool __maybe_unused gpio_is_range_overlap(unsigned int start,
+						 unsigned int end)
+{
+	struct bcm_gpio_chip *gc = NULL;
+
+	SLIST_FOREACH(gc, &gclist, link)
+		if ((start < (gc->gpio_base + gc->ngpios)) &&
```

Comment:
not quite sure here, `||` instead of `&&`?

Reply:
The formula detects if there is an overlap and stops traversing rest of the list by returning `true`.
You might be seeing w.r.t detect non-overlapping range.

Reply:
Assuming the number of gpios in a block is always the same, then `start < (gc->gpio_base + gc->ngpios)` means there's an overlap with a lower numbered block.
`end > gc->gpio_base` means there's an overlap with a higher numbered block.
Right? If so, then both conditions can't be true at the same time and will always `return false`.
The only time it can `return true` is if you pass in a super block that has so many of gpios that it overlaps both ends of the current block.
Do all the blocks not have the same number of gpios?

Reply:
It is independent of number of GPIOs in a block.  GPIO ID passed by each block must be unique.
ex:
GPIO chip 0: base 32  ngpios 32   pin id range: `32 ~ 63 ` 
GPIO chip 1:  ranges `0~31,  64~95` or beyond are OK.
Anything intersecting range `32~63` is overlap.
`if (start < 64  &&  end  > 32)  overlap.`

Reply:
Ok, got it. Thanks!